### ISSUE TITLE
Duplicate votes being recorded in a block

### ISSUE THREAD

**Issue by mattcollier** [digitalbazaar/bedrock-ledger-consensus-continuity#44]
This issue was discovered while doing integration testing in `bedrock-ledger-agent`, see an example test run on CI: https://ci.digitalbazaar.com/job/bedrock-ledger-agent/40/console

This issue only manifests on the CI system (not on my local system) which likely means it's a race condition.

**Comment by mattcollier** [comment]
This gist shows:

last two votes are from voter ...9476e02, the second is for `voteRound` 2, all other votes are for `voteRound` 1

https://gist.github.com/mattcollier/daa13feaeb58c6e790f26974f0ffd930#file-duplicatevotes-json-L83-L136

**Comment by mattcollier** [comment]
I'm not sure if this is the root cause or not, but...

https://github.com/digitalbazaar/bedrock-ledger-consensus-continuity/blob/master/lib/server.js#L285

The call to `storage.votes.get` there, does not do any filtering related to the `voteRound` of the vote.  So, if a voter has moved on to doing a second round of voting, their vote for the first *and* second round will be included here, ultimately leading to those votes being included in the consensus block.

This CI test run includes logging that demonstrates this issue: https://ci.digitalbazaar.com/job/bedrock-ledger-agent/63/console

**Comment by dlongley** [comment]
We should definitely only include the first successful round -- so whatever filter is necessary to make that happen should be added. I'm *pretty sure* it's ok for a node to proceed with a second round of voting when there is a delay in the votes for the first round but my brain has switched contexts so I'd have to go back and look to ensure that's not also a bug. Either way, only first successful round should go into the block.

**Comment by dlongley** [comment]
Fixing up where we get the votes in the server code (that code you linked to) should only really be an optimization, I think -- and I'm not sure if it's even valid. The real problem should be where the block gets created. The votes that go into it should only be from the round that was successful.

**Comment by dlongley** [comment]
Specifically here: https://github.com/digitalbazaar/bedrock-ledger-consensus-continuity/blob/master/lib/worker.js#L506

**Comment by dlongley** [comment]
Hmm, I see that the votes there are filtered based on the roll-call manifest:

https://github.com/digitalbazaar/bedrock-ledger-consensus-continuity/blob/master/lib/worker.js#L510-L512

Which is correct. So it must be that there's a bug with generating the roll-call manifest -- where it's including votes beyond the successful round.

**Comment by dlongley** [comment]
I think the bug is here:

https://github.com/digitalbazaar/bedrock-ledger-consensus-continuity/blob/master/lib/election.js#L386-L387

It fetches *all* vote hashes instead of just those up to the successful vote round.

**Comment by mattcollier** [comment]
@dlongley the server routine and the writeBlock routine are both calling the same `get` API: https://github.com/digitalbazaar/bedrock-ledger-consensus-continuity/blob/master/lib/server.js#L285

https://github.com/digitalbazaar/bedrock-ledger-consensus-continuity/blob/master/lib/worker.js#L506

**Comment by dlongley** [comment]
We need two fixes:

1. Fix the code I linked to above so that we only include votes up to the winning round in our roll call manifest.
2. Fix the code that checks the roll-call manifest to make sure it doesn't include votes beyond the winning around.

**Comment by mattcollier** [comment]
@dlongley I have modified the `election.tally` API as we discussed.  See: https://github.com/digitalbazaar/bedrock-ledger-consensus-continuity/compare/fix44
The test suite within consensus-continuity still passes with this change.

Were you imagining that the `election.tally` API would be called from within [_createRollCallManifest](https://github.com/digitalbazaar/bedrock-ledger-consensus-continuity/blob/fix44/lib/election.js#L615) and the `election.tally().round` would be used to filter the votes (and by extension vote hashes) there?

That *is* what I was imagining, but at the point in time when `_createRollCallManifest` is called, `election.tally` *always* returns null, because a majority of votes have not been collected, and therefore there is no winner yet.
